contributors/design-proposals/auth/self-hosting-authorization-webhook.md

```diff
@@ -0,0 +1,369 @@
+      client-key: /path/to/key.pem          # key matching the cert
+
+# kubeconfig files require a context. Provide one for the API Server.
+current-context: webhook
+contexts:
+- context:
+    cluster: name-of-remote-authz-service
+    user: name-of-api-server
+  name: webhook
+```
+
+The kubeconfig file format does not easily admit reference to an endpoint that
+is hosted within the cluster.  Earlier proposals to extend the syntax of the
+`clusters[*].cluster.server` field to admit a custom dialer that calls a service
+endpoint [were rejected][4] (also see [another rationale][5]).  The [suggested
+viable alternative][6] is to provide a new webhook file format that references
+a service explicitly, and build on top of that.
+
+One further concern to address is that this proposed change is only [relevant
+for the `kube-apiserver`][7], and not to any of the generic apiservers.  This
+means that the proposed changes will be set up such that only the
+`kube-apiserver` can make use of it.  This approach has [garnered support][8]
+from the community in the earlier proposal.
+
+## Proposal
+
+* Define a new configuration format, supported only by `kube-apiserver` that
+  admits a webhook endpoint within the cluster to be specified.
+
+  This new configuration file is still specified using the same flag as for the
+  original webhook configuration, `--auhtorization-webhook-config-file`.
+
+* Instantiate a custom subject access review client based on this configuration
+  that has special behavior as follows:
+
+  - If the new configuration has been specified, a custom dialer is supplied
+    which resolves a service endpoint and uses that service endpoint to dial
+    into when making a call.
+
+## User Experience
+
+### Use Cases
+
+1. As a cluster administrator, I would like to install Kubernetes-based products
+   that have self-contained deployments in a cluster and include authorization.
+
+   Kubernetes products typically come with an install script, say based on Helm
+   or `kubeadm`.  These allow custom deployments, including reconfiguring
+   `kube-apiserver`.  When such a product requires an authorizer, the authorizer
+   is added as a system component with a pod-based deployment in a cluster and
+   the apiserver is instructed to consult the endpoint based on that deployment
+   for authorization decisions.
+
+   The cluster administrator experience changes in that they are able to install
+   a wider range of Kubernetes-based products that include this feature,
+   compared to when this is not the case.
+
+## Implementation
+
+
+### Configuration format
+
+Let us define a new configuration file format.  This is based on the [admission
+configuration file format][12] as well as the proposal from @frankfarzan for an
+analogous [configuration format for authz][13].
+
+An example configuration file for authz looks like this:
+
+```yaml
+apiVersion: authorizationconfig.k8s.io/v1alpha1
```

Comment:
I know we've had concerns in the past about API groups with only one type. In the future, should this group hold other webhook configs? e.g. audit or authentication?

Reply:
What do you recommend to do here instead?

Not sure I can decide either way, not knowing the history of such concerns.

Reply:
This was referencing https://github.com/kubernetes/kubernetes/pull/55812#discussion_r151370163

Maybe this could be included in the existing `apiserver.config.k8s.io` group?

https://github.com/kubernetes/kubernetes/pull/55812/commits/229c4307b173842633327a4fea3f16cde73cf196#diff-77d495a77d16f1f8c941606e2e5dee42R30

Reply:
My understanding from previous comments is that this addition would be a kube-apiserver feature only, and as such not suitable for generic apiservers.

See for example discussion around this comment: https://github.com/kubernetes/kubernetes/issues/52511#issuecomment-333224542

Would your proposal not mean that we're mixing in these specific concerns into all other apiservers?

Reply:
Didn't realized deads2k had already commented on this. It's fine as is then. Thanks for the feedback.

Reply:
Sure thing!  It's on me as a proposer to dig up prior art.

Reply:
Per https://github.com/kubernetes/community/blob/master/contributors/devel/component-config-conventions.md, component config apigroups should have the prefix: `.config.k8s.io`. I also don't see why this specific config format will not be useful for other webhooks (e.g. authenticator).

Reply:
Changed the prefix to conform to that convention if it helps.  I'm not generalizing since this is not the intention, and previous feedback has been that this should remain a kubeapiserver feature only.